File: NeetCode/Graph/Island_Perimeter/island__perimeter.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        height = len(grid)
        width = len(grid[0])

        perimeter = 0

        for y in range(height):
            for x in range(width):
                if grid[y][x] == 0:
                    continue

                perimeter += 4

                if y > 0 and grid[y - 1][x] == 1:
                    perimeter -= 2

                if x > 0 and grid[y][x - 1] == 1:
                    perimeter -= 2
        return perimeter


if __name__ == "__main__":
    sol = Solution()
    grid = [
        [1, 1, 0, 0],
        [1, 0, 0, 0],
        [1, 1, 1, 0],
        [0, 0, 1, 1],
    ]
    print(sol.islandPerimeter(grid))
    print("Running Solution...")


""" 
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        diff = ((0, 1), (0, -1), (1, 0), (-1, 0))
        height = len(grid)
        width = len(grid[0])
        inUse = set()

        def dfs(y, x):
            if y < 0 or y == height or x < 0 or x == width or (y, x) in inUse:
                return 0

            total = 4
            inUse.add((y, x))
            for dy, dx in diff:

                ny, nx = dy + y, dx + x
                if 0 <= ny < height and 0 <= nx < width and grid[ny][nx] == 1:
                    total -= 1
                    total += dfs(ny, nx)

            return total

        for y in range(height):
            for x in range(width):
                if grid[y][x] == 1:
                    return dfs(y, x)
        return 0
"""

File: NeetCode/Graph/Island_Perimeter/Island_Perimeter.cpp
// Island_Perimeter.cpp
//
// Compile locally with:
// clang++ Island_Perimeter.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int islandPerimeter(vector<vector<int>> &grid)
    {
        int height = grid.size();
        int width = grid[0].size();

        int perimeter = 0;

        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                if (grid[y][x] == 0)
                {
                    continue;
                }
                perimeter += 4;
                if (y > 0 && grid[y - 1][x] == 1)
                {
                    perimeter = perimeter - 2;
                }
                if (x > 0 && grid[y][x - 1] == 1)
                {
                    perimeter = perimeter - 2;
                }
            }
        }
        return perimeter;
    }
};

int main()
{
    cout << "Running Island_Perimeter.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
